File: python/day06-errors-modules/solutions.py
"""Day 6 reference solutions."""

import math


class NegativeAmountError(Exception):
    pass


class InsufficientFundsError(Exception):
    pass


def safe_divide(a, b):
    try:
        return a / b
    except ZeroDivisionError:
        return None


def parse_int_or_default(value, default=0):
    try:
        return int(value)
    except ValueError:
        return default


def withdraw(balance, amount):
    if amount < 0:
        raise NegativeAmountError(f"amount cannot be negative: {amount}")
    if amount > balance:
        raise InsufficientFundsError(f"need {amount}, have {balance}")
    return balance - amount


def sqrt_of_sum_of_squares(*numbers):
    return math.sqrt(sum(n * n for n in numbers))


def main():
    print("Running Day 6 exercises directly.")

File: typescript-javascript/day06-errors-modules/solutions.js
// Day 6 reference solutions.

class NegativeAmountError extends Error {
    constructor(message) {
        super(message);
        this.name = "NegativeAmountError";
    }
}

class InsufficientFundsError extends Error {
    constructor(message) {
        super(message);
        this.name = "InsufficientFundsError";
    }
}

function safeDivide(a, b) {
    if (b === 0) {
        return null;
    }
    return a / b;
}

function parseIntOrDefault(value, defaultValue = 0) {
    const parsed = parseInt(value, 10);
    if (Number.isNaN(parsed)) {
        return defaultValue;
    }
    return parsed;
}

function withdraw(balance, amount) {
    if (amount < 0) {
        throw new NegativeAmountError(`amount cannot be negative: ${amount}`);
    }
    if (amount > balance) {
        throw new InsufficientFundsError(`need ${amount}, have ${balance}`);
    }
    return balance - amount;
}

module.exports = { NegativeAmountError, InsufficientFundsError, safeDivide, parseIntOrDefault, withdraw };
